fix(build-tools): handle special export cases

jason-ha opened the pull request:
- Look for namespace exports and source local exports that may have JSDocs that are determined using ExportedDeclarations.
- Export all unknown matching api-extractor.
- Change unknown case to be a map and additional data as collected.
- Don't export anything when there are none (otherwise it exported all)

- Fix incrementality task typo

jason-ha committed:
fix(build-tools): handle special export cases
- Look for namespace exports and source local exports that may have JSDocs that are determined using ExportedDeclarations.

- Change unknown case to be a map and additional data as collected.

- Fix incrementality task typo

Josmithr commented on build-tools/packages/build-cli/src/library/typescriptApi.ts:
```diff
@@ -99,5 +117,73 @@ export function getApiExports(sourceFile: SourceFile): ExportRecords {
+	// given file's export declarations.
+	const exportDeclarations = sourceFile.getExportDeclarations();
+	for (const exportDeclaration of exportDeclarations) {
+		// console.log(`export @ ${exportDeclaration.getStartLineNumber()} has children:\n\t${exportDeclaration.getChildren().map((c) => c.getKindName()).join('\n\t')}`);
```
Remove?

jason-ha replied:
Nah - it is good for debugging - I was going to comment it though :)

Josmithr commented on build-tools/packages/build-cli/src/library/typescriptApi.ts:
```diff
@@ -99,5 +117,73 @@ export function getApiExports(sourceFile: SourceFile): ExportRecords {
+	for (const exportDeclaration of exportDeclarations) {
+		// console.log(`export @ ${exportDeclaration.getStartLineNumber()} has children:\n\t${exportDeclaration.getChildren().map((c) => c.getKindName()).join('\n\t')}`);
+
+		// Look first for namespace exports like `export * as foo`.
```
Presumably this covers explicit `namespace` declarations as well?

jason-ha replied:
If a namespace is exported that should be covered by the logic below. If you have specific cases we can try 'em, out.
Tests are sorely lacking - especially when the results are not checked in.

Josmithr commented on build-tools/packages/build-cli/src/library/typescriptApi.ts:
```diff
@@ -99,5 +117,73 @@ export function getApiExports(sourceFile: SourceFile): ExportRecords {
+		//     /** @internal */
+		//     foo
+		//   }
+		// Note that per namespace level tags maybe different than the namespace. To avoid any
```
I suspect this is why API-Extractor hasn't implemented this yet. In theory, we need to recurse into the namespace's child items and do the appropriate trimming at that layer. I think namespaces are a case where redeclaration is necessary, and I think it's safe to do so. 

I wonder if we could at least add some validation here that there aren't any items being included that shouldn't. How easy would this be?

jason-ha replied:
Presumable you should be able to enumerate all of the things inside the namespace/object being exported.
It would probably be better to have multiple namespaces. Redeclaring is where we get into trouble. These files need to export exactly what is in the original because that is what will be there at runtime.
It would be okay to do something like type manipulation to Omit members - don't know if such a thing is possible.

jason-ha committed:
docs(build-tools): improved comments

Josmithr commented on build-tools/packages/build-cli/src/library/typescriptApi.ts:
```diff
@@ -35,29 +35,47 @@ function isTypeExport(_decl: ExportedDeclarations): boolean {
-		if (parent !== undefined) {
-			return getNodeLevel(parent);
-		}
+		return getLevelFromDocs(node.getJsDocs());
```
It just occurred to me - does this analyze / trim based on the tags of items being re-exported from other packages? That's useful for our own packages, but could be problematic if we happen to re-export from some external library that happens to use TSDoc release tags.

Likely not something we _really_ need to be worried about, but could occur in theory, and is something API-Extractor has to contend with (albeit, not well, currently). Might be worth a note though.

jason-ha replied:
This logic should let us retag things, but I didn't really get into the details.
I do know that we have one type that we surface from another library - `EventEmitter`. I know because the rewriter logic complains about it not having a tag when re-exported by example-utils.

jason-ha committed:
fix(build-tools): fix some entrypoints cases
- export all unknown matching api-extractor
- skip export when nothing to export (instead of exporting everything)

tylerbutler commented:
Since Jason is out, I'll merge this once @Josmithr has had a chance to review again.

tylerbutler commented:
Looks like both Jason and Josh are out, so merging.
